src/Quartz/Simpl/TaskSchedulingThreadPool.cs

```diff
@@ -212,44 +227,47 @@ public bool RunInThread(Func<Task> runnable)
         }
 
         /// <summary>
-        /// Removes a task from the 'running' list (if it exists there) and releases
-        /// the concurrency semaphore so that more tasks may begin running
+        /// Decrements the number of running tasks and releases the concurrency semaphore so that more
+        /// tasks may begin running.
         /// </summary>
         /// <param name="completedTask">The task which has completed</param>
         private void RemoveTaskFromRunningList(Task completedTask)
         {
             concurrencySemaphore.Release();
-            lock (taskListLock)
-            {
-                runningTasks.Remove(completedTask);
-            }
+            runningTasksCountdown.Signal();
         }
 
         /// <summary>
-        /// Stops processing new tasks and optionally waits for currently running tasks to finish
+        /// Stops processing new tasks and optionally waits for currently running tasks to finish.
         /// </summary>
-        /// <param name="waitForJobsToComplete">True to wait for currently executing tasks to finish; false otherwise</param>
+        /// <param name="waitForJobsToComplete"><see langword="true"/> to wait for currently executing tasks to finish; otherwise, <see langword="false"/>.</param>
         public void Shutdown(bool waitForJobsToComplete = true)
         {
             log.Debug("Shutting down threadpool...");
 
             // Cancel using our shutdown token
             shutdownCancellation.Cancel();
 
-            // If waitForJobsToComplete is true, wait for runningTasks
+            // If waitForJobsToComplete is true, wait for running tasks to complete
             if (waitForJobsToComplete)
             {
-                Task[] tasksArray;
-                lock (taskListLock)
+                lock (runningTasksCountdown)
                 {
                     // Cancellation has been signaled, so no new tasks will begin once
                     // shutdown has acquired this lock
-                    tasksArray = runningTasks.ToArray();
+                    log.DebugFormat($"Waiting for {runningTasksCountdown.CurrentCount.ToString()} threads to complete.");
```

Comment:
This is a trap that a lot of people run into. String interpolation, just like `string.Format(...)` will box (struct) parameters.